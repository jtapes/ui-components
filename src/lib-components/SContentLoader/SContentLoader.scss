.s-content-loader {
  display: flex;
  font-size: 0;

  svg {
    width: auto;
    height: auto;
  }
}
