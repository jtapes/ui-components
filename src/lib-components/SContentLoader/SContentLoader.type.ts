export type SContentLoaderType = string | number
