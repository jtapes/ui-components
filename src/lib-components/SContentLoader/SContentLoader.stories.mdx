import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Canvas,
} from '@storybook/addon-docs'
import SContentLoader from '@/lib-components/SContentLoader/SContentLoader.vue'
import results from '/.jest-test-results.json';

<Meta
  title="DESIGN SYSTEM/Atom/ContentLoader"
  component={SContentLoader}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    width: { number: 'number'},
    height: { number: 'number'},
    primaryColor: {control: 'text', defaultValue: '#f4f4F4'},
    secondaryColor: {control: 'text', defaultValue: '#d9e2ec'},
    primaryOpacity: {number: 'number', defaultValue: 1},
    secondaryOpacity: {number: 'number', defaultValue: 0},
    speed: {number: 'number'},
    baseUrl: {control: 'text'},
    animate: {control: 'boolean', defaultValue: true},
    uniqueKey: {control: 'text'}
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/IIeFkxNoHGcaqNHlVeNrhW/MVP-3-Market-UI-Kit?node-id=846%3A1100',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# SContentLoader

<Description of={SContentLoader} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SContentLoader },
  template:
    `<SContentLoader :width="width" :height="height" :primaryColor="primaryColor" :secondaryColor="secondaryColor" :primaryOpacity="primaryOpacity"
:secondaryOpacity="secondaryOpacity" :speed="speed" :animate="animate" :uniqueKey="uniqueKey"></SContentLoader>`
})

export const TemplateChips = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SContentLoader },
  template:
    `<SContentLoader :width="width" :height="height" :primaryColor="primaryColor" :secondaryColor="secondaryColor"
:primaryOpacity="primaryOpacity" :secondaryOpacity="secondaryOpacity" :speed="speed" :animate="animate" :uniqueKey="uniqueKey">
      <rect x="0" y="0" rx="24" ry="24" width="200" height="48" />
      <rect x="212" y="0" rx="24" ry="24" width="200" height="48" />
      <rect x="424" y="0" rx="24" ry="24" width="200" height="48" />
      <rect x="636" y="0" rx="24" ry="24" width="200" height="48" />
    </SContentLoader>`
})

<Canvas withToolbar withSource="none">
  <Story name="SContentLoader" args={{  }}>
    {Template.bind({})}
  </Story>
  <Story name="SContentLoaderChips" args={{  }}>
    {TemplateChips.bind({})}
  </Story>
</Canvas>

## Дефолтный пример

<Source
  language="html"
  code={`
    <s-content-loader />
  `}
/>

## Пример для тегов

<Source
  language="html"
  code={`
    <s-content-loader width="850" height="48">
      <rect x="0" y="0" rx="24" ry="24" width="200" height="48" />
      <rect x="212" y="0" rx="24" ry="24" width="200" height="48" />
      <rect x="424" y="0" rx="24" ry="24" width="200" height="48" />
      <rect x="636" y="0" rx="24" ry="24" width="200" height="48" />
    </s-content-loader>
  `}
/>
<ArgsTable of={SContentLoader} />
