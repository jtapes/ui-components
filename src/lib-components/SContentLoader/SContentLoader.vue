<template>
  <div class="s-content-loader">
    <svg
      :width="width"
      :height="height"
      :viewBox=" `0 0 ${width} ${height}`"
      version="1.1"
      :preserveAspectRatio="preserveAspectRatio"
    >
      <rect
        :style="{ fill: `url(${baseUrl}#${idGradient})` }"
        :clip-path="`url(${baseUrl}#${idClip})`"
        x="0"
        y="0"
        width="100%"
        height="100%"
      />

      <defs>
        <clipPath
          :id="idClip"
        >
          <!--  @slot Слот для контента  -->
          <slot>
            <rect
              x="0"
              y="0"
              rx="5"
              ry="5"
              width="100%"
              height="100%"
            />
          </slot>
        </clipPath>

        <linearGradient :id="idGradient">
          <stop
            offset="0%"
            :stop-color="primaryColor"
            :stop-opacity="primaryOpacity"
          >
            <animate
              v-if="animate"
              attributeName="offset"
              values="-2; 1"
              :dur="`${speed}s`"
              repeatCount="indefinite"
            />
          </stop>
          <stop
            offset="50%"
            :stop-color="secondaryColor"
            :stop-opacity="secondaryOpacity"
          >
            <animate
              v-if="animate"
              attributeName="offset"
              values="-1.5; 1.5"
              :dur="`${speed}s`"
              repeatCount="indefinite"
            />
          </stop>
          <stop
            offset="100%"
            :stop-color="primaryColor"
            :stop-opacity="primaryOpacity"
          >
            <animate
              v-if="animate"
              attributeName="offset"
              values="-1; 2"
              :dur="`${speed}s`"
              repeatCount="indefinite"
            />
          </stop>
        </linearGradient>
      </defs>
    </svg>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import { SContentLoaderType } from '@/lib-components/SContentLoader/SContentLoader.type'

/** Компонент для отображения загрузки контента */
@Component
export default class SContentLoader extends Vue {
  /** Высота */
  @Prop({ type: [String,Number], default: 200 })
  height!: SContentLoaderType

  /** Ширина */
  @Prop({ type: [String,Number], default: 200 })
  width!: SContentLoaderType

  /** Основной цвет */
  @Prop({ type: [String,Number], default: '#f4f4F4' })
  primaryColor!: SContentLoaderType

  /** Второй цвет */
  @Prop({ type: String, default: '#d9e2ec' })
  secondaryColor!: string

  /** Прозрачность основного цвета */
  @Prop({ type: [String,Number], default: 1 })
  primaryOpacity!: SContentLoaderType

  /** Прозрачность второго цвета */
  @Prop({ type: [String,Number], default: 0 })
  secondaryOpacity!: SContentLoaderType

  /** Скорость анимации - чем больше значение, тем скорость медленнее */
  @Prop({ type: [String,Number], default: 2 })
  speed!: SContentLoaderType

  /** Url для маски свг */
  @Prop({ type: String, default: '' })
  baseUrl!: string

  /** Включение анимации */
  @Prop({ type: Boolean, default: true })
  animate!: boolean

  /** Уникальный идентификатор */
  @Prop({ type: [String,Number] })
  uniqueKey?: string | number

  @Prop({ type: String, default: 'xMinYMin slice' })
  preserveAspectRatio!: string

  /** Генерация уникальных id */
  idGradient = this.uniqueKey
    ? `${this.uniqueKey}-idGradient`
    : Math.random().toString(36).substring(2)
  idClip = this.uniqueKey
    ? `${this.uniqueKey}-idClip`
    : Math.random().toString(36).substring(2)

}
</script>
