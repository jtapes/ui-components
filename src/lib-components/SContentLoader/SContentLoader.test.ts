import { mount } from '@vue/test-utils'
import SContentLoader from '@/lib-components/SContentLoader/SContentLoader.vue'

describe('SContentLoader', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SContentLoader, {
      propsData: {
        width: '100%',
        height: '100%',
        primaryColor: '#f4f4F4',
        secondaryColor: '#d9e2ec',
        primaryOpacity: 1,
        secondaryOpacity: 0,
        speed: 1,
        baseUrl: 'loader',
        animate: true,
        uniqueKey: 'uniqueKey-content'
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
