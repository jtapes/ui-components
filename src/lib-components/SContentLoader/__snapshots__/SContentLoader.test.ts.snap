// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SContentLoader Is Vue snapshot 1`] = `
<div
  class="s-content-loader"
>
  <svg
    height="100%"
    preserveAspectRatio="xMinYMin slice"
    version="1.1"
    viewBox="0 0 100% 100%"
    width="100%"
  >
    <rect
      clip-path="url(loader#uniqueKey-content-idClip)"
      height="100%"
      style="fill: url(loader#uniqueKey-content-idGradient);"
      width="100%"
      x="0"
      y="0"
    />
     
    <defs>
      <clippath
        id="uniqueKey-content-idClip"
      >
        <rect
          height="100%"
          rx="5"
          ry="5"
          width="100%"
          x="0"
          y="0"
        />
      </clippath>
       
      <lineargradient
        id="uniqueKey-content-idGradient"
      >
        <stop
          offset="0%"
          stop-color="#f4f4F4"
          stop-opacity="1"
        >
          <animate
            attributeName="offset"
            dur="1s"
            repeatCount="indefinite"
            values="-2; 1"
          />
        </stop>
         
        <stop
          offset="50%"
          stop-color="#d9e2ec"
          stop-opacity="0"
        >
          <animate
            attributeName="offset"
            dur="1s"
            repeatCount="indefinite"
            values="-1.5; 1.5"
          />
        </stop>
         
        <stop
          offset="100%"
          stop-color="#f4f4F4"
          stop-opacity="1"
        >
          <animate
            attributeName="offset"
            dur="1s"
            repeatCount="indefinite"
            values="-1; 2"
          />
        </stop>
      </lineargradient>
    </defs>
  </svg>
</div>
`;
